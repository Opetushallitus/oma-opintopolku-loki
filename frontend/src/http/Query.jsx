import React from 'react'
import PropTypes from 'prop-types'
import http from './http'

class Query extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      data: null,
      error: null,
      pending: true
    }
  }

  componentDidMount () {
    this.doQuery()
  }

  async doQuery () {
    const { url, method, body } = this.props

    try {
      const { data } = await http.request(url, method, body)
      this.setState({ data, pending: false })
    } catch (error) {
      console.error(error)
      this.setState({ error, pending: false })
    }
  }

  render () {
    return this.props.children(this.state)
  }
}

Query.defaultProps = {
  method: 'get',
  body: {}
}

Query.propTypes = {
  url: PropTypes.string.isRequired,
  method: PropTypes.string,
  body: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array
  ]),
  children: PropTypes.func.isRequired
}

export default Query
