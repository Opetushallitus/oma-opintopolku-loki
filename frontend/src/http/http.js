import axios from 'axios'

const client = axios.create({
  baseURL: process.env.API_BASE_URL
})

const get = async url => client.get(url)
const post = async (url, data) => client.post(url, data)

const request = async (url, method, data = {}) => {
  switch (method) {
    case 'GET':
    case 'get':
      return get(url)
    case 'POST':
    case 'post':
      return post(url, data)
    default:
      throw new Error(`Unsupported HTTP method: ${method}`)
  }
}

export default {
  get,
  post,
  request
}
