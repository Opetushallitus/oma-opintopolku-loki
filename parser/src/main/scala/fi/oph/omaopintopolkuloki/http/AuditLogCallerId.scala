package fi.oph.omaopintopolkuloki.http

object AuditLogCallerId {
  val code = "1.2.246.562.10.00000000001.oma-opintopolku-loki/parser"
}
