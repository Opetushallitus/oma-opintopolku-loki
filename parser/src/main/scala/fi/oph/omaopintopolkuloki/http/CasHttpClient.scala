package fi.oph.omaopintopolkuloki.http

import fi.oph.omaopintopolkuloki.conf.Configuration
import fi.vm.sade.utils.cas.{CasAuthenticatingClient, CasClient, CasParams}
import org.http4s.client._


object CasHttpClient {

  val sessionCookieName = "JSESSIONID"

  def apply(httpClient: Client, scheme_authority: String, params: CasParams = Params.permission) = {
    CasAuthenticatingClient(
      new CasClient(scheme_authority, httpClient),
      params,
      httpClient,
      Some(AuditLogParserSubsystemCode.code),
      sessionCookieName
    )
  }
}


object Params {
  private lazy val permission_path = "/kayttooikeus-service"

  private lazy val credentials = Configuration.getSecrets

  def permission: CasParams = CasParams(permission_path, credentials.username, credentials.password)
}

