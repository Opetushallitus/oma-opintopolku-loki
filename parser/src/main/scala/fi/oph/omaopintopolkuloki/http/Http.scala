package fi.oph.omaopintopolkuloki.http

import cats.effect.IO
import cats.effect.unsafe.IORuntime
import fi.oph.omaopintopolkuloki.conf.Configuration._
import org.http4s.blaze.client.BlazeClientBuilder
import org.http4s.client.Client
import org.http4s.{Request, Uri}

import scala.concurrent.ExecutionContext.global

trait HttpClient {
  type Decode[ResultType] = (Int, String, Request[IO]) => ResultType

  def get[ResultType](uri: Uri)(decode: Decode[ResultType]): IO[ResultType]
}

object Http {
  implicit private val runtime: IORuntime = cats.effect.unsafe.IORuntime.global

  def apply(useCas: Boolean = false): HttpClient = {
    val client = BlazeClientBuilder[IO](global)
      .withMaxTotalConnections(maxRequestThreads)
      .allocated
      .map(_._1)
      .unsafeRunSync()

    if (useCas) {
      new Http(CasHttpClient(client, baseURI.renderString + "/cas"))
    } else {
      new Http(client)
    }
  }
}

private class Http(client: Client[IO]) extends HttpClient {

  def get[ResultType](uri: Uri)(decode: Decode[ResultType]): IO[ResultType] = {
    processRequest(Request[IO](uri = uri))(decode)
  }

  private def processRequest[ResultType](request: Request[IO])(decoder: (Int, String, Request[IO]) => ResultType): IO[ResultType] = {
    client.run(request).use { response =>
      response.as[String].map(text => decoder(response.status.code, text, request))
    }
  }
}
