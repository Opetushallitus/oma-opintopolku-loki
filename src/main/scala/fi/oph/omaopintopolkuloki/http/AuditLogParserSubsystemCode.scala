package fi.oph.omaopintopolkuloki.http

object AuditLogParserSubsystemCode {
  val code = "auditlog"
}
