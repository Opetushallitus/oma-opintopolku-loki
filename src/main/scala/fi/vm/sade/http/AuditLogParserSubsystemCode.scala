package fi.vm.sade.http

object AuditLogParserSubsystemCode {
  val code = "auditlog"
}
