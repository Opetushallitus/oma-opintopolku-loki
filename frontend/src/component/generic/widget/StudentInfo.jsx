import React from 'react'
import styled from 'styled-components'
import { H3 } from 'ui/typography'
import { parseUserName } from 'service/raamitSupport'
import Query from 'http/Query'

const Student = styled(H3)`
  margin-top: 3rem;
  margin-bottom: 2.214rem;
`

export const Name = styled.span`
  font-weight: 600;
`
export const Birthday = styled.span`
  font-weight: 400;
`

const parseCentury = (hetu) => {
  const pattern = /\d{6}([-A+])/
  const matches = hetu.match(pattern)
  const separator = matches[1]

  switch (separator) {
    case '+':
      return 1800
    case separator.match(/[-YXWVU]/)?.input: // hack to use regex in switch statement
      return 1900
    case separator.match(/[ABCDEF]/)?.input:
      return 2000
  }
}

const parseBirthday = (hetu) => {
  try {
    const pattern = /(\d{2})/g
    const matches = hetu.match(pattern)

    const day = parseInt(matches[0])
    const month = parseInt(matches[1])
    const yearSuffix = parseInt(matches[2])

    const year = parseCentury(hetu) + yearSuffix

    return `${day}.${month}.${year}`
  } catch (err) {
    console.log('Failed to parse birthday', err) // TODO: Log errors remotely
  }
}

const StudentInfo = () => (
  <Query url='whoami'>
    {({ data, error, pending }) => {
      if (error || pending) return null

      const name = parseUserName(data)
      const birthday = parseBirthday(data.hetu)

      return (
        <Student>
          <Name>{name}</Name>
          { birthday ? <Birthday> s. {birthday}</Birthday> : null }
        </Student>
      )
    }}
  </Query>
)

export default StudentInfo
