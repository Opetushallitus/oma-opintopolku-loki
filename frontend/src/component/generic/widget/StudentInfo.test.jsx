import React from 'react'
import StudentInfo, { Birthday, Name } from './StudentInfo'
import { create } from 'react-test-renderer'
import http from '../../../http/http'

describe('StudentInfo', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('should be rendered', (done) => {
    http.request = jest.fn(() =>
      Promise.resolve({
        data: {
          etunimet: 'Onni Oskari',
          sukunimi: 'Oppija',
          hetu: '010280-ABC1'
        }
      })
    )

    const renderer = create(<StudentInfo/>)

    setTimeout(() => {
      expect(renderer.root.findByType(Name).props.children).toEqual('Onni Oskari Oppija')
      expect(renderer.root.findByType(Birthday).props.children.join('')).toEqual(' s. 1.2.1980')

      const componentJSON = renderer.toJSON()
      expect(componentJSON).toMatchSnapshot()
      renderer.unmount()

      expect(http.request).toHaveBeenCalledTimes(1)

      done()
    })
  })

  it('parses birthdate correctly', (done) => {
    http.request = jest.fn(() =>
      Promise.resolve({
        data: {
          etunimet: 'Onni Oskari',
          sukunimi: 'Oppija',
          hetu: '010203AABC1'
        }
      })
    )

    const renderer = create(<StudentInfo/>)

    setTimeout(() => {
      expect(renderer.root.findByType(Name).props.children).toEqual('Onni Oskari Oppija')
      expect(renderer.root.findByType(Birthday).props.children.join('')).toEqual(' s. 1.2.2003')
      expect(renderer.root.findAllByType('span')).toHaveLength(2)

      renderer.unmount()

      expect(http.request).toHaveBeenCalledTimes(1)

      done()
    })
  })

  it('parses really old birthdate correctly', (done) => {
    http.request = jest.fn(() =>
      Promise.resolve({
        data: {
          etunimet: 'Onni Oskari',
          sukunimi: 'Oppija',
          hetu: '010299+ABC1'
        }
      })
    )

    const renderer = create(<StudentInfo/>)

    setTimeout(() => {
      expect(renderer.root.findByType(Name).props.children).toEqual('Onni Oskari Oppija')
      expect(renderer.root.findByType(Birthday).props.children.join('')).toEqual(' s. 1.2.1899')
      expect(renderer.root.findAllByType('span')).toHaveLength(2)

      renderer.unmount()

      expect(http.request).toHaveBeenCalledTimes(1)

      done()
    })
  })

  it('handles invalid hetu correctly', (done) => {
    http.request = jest.fn(() =>
      Promise.resolve({
        data: {
          etunimet: 'Onni Oskari',
          sukunimi: 'Oppija',
          hetu: 'foobar'
        }
      })
    )
    console.log = jest.fn()

    const renderer = create(<StudentInfo/>)

    setTimeout(() => {
      expect(renderer.root.findByType(Name).props.children).toEqual('Onni Oskari Oppija')
      expect(renderer.root.findAllByType('span')).toHaveLength(1)

      renderer.unmount()

      expect(console.log).toHaveBeenCalledTimes(3)

      done()
    })
  })
})
