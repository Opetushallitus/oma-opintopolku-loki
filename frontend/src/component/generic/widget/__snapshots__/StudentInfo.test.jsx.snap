// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentInfo should be rendered 1`] = `
[
  <h3
    className="typography__H3-sc-1yq3b4j-2 StudentInfo__Student-sc-8zvgal-0 dJnBdy cpuvVu"
  >
    <span
      className="StudentInfo__Name-sc-8zvgal-1 eOXinT"
    >
      Onni Oskari Oppija
    </span>
    <span
      className="StudentInfo__Birthday-sc-8zvgal-2 cDqyaU"
    >
       s. 
      1.2.1980
    </span>
  </h3>,
  <div>
    <b
      className="typography__Bold-sc-1yq3b4j-6 jFpkKS"
    >
      Tarkasteltava henkil√∂:
    </b>
    <select
      className="typography__Select-sc-1yq3b4j-7 jmTUoG"
      onChange={[Function]}
    >
      <option
        value="010280-ABC1"
      >
        Onni Oskari Oppija
         
         s. 1.2.1980
      </option>
    </select>
  </div>,
]
`;
